IDEMPIERE-5013 Implement HikariCP as a replacement for c3p0

siepkes opened the pull request:
Hi all!

I order to support PostgreSQL reconnecting after failover when using client side node selection I replaced C3P0 with HikariCP. HikariCP is an actively maintained, Apache licensed connection pool with substantially better performance and better resilience to failure (DB disconnects, etc.) then C3P0. Read more about it here: https://github.com/brettwooldridge/HikariCP (including comparison and benchmarks with C3P0). During the development of HikariCP a lot of care was spend on making HikariCP behave as correctly and predictable as possible.

Let me know what you think!

Aside from HikariCP the following changes were made:

Cleaned up the `getCachedConnection` method. With HikariCP there is no need to retry to obtain a connection since getting an connection will block until a free connection is available or until a timeout is reached (default 30 seconds) at which point an `SQLException` is thrown. This also removed calling `Runtime.getRuntime().runFinalization();`. HikariCP is currently configured to detect / log leaks when a connection hasn't returned to the pool for longer then 5 minutes.

Initialization of the datasource must happen in the `getDataSource()` method because at object construction not all JDBC config is known. However this method could (as far as I could tell) be called concurrently from multiple threads but had no mechanism to prevent initializing the DB pool multiple times. The variable in which the pool itself was stored (`m_ds`) also was not marked volatile or immutable which could lead to visibility issues. Instead of lazy initialization of the pool in the `getDataSource()` method the pool could probably better be initialized at object construction. However I wasn't able to achieve that without breakage therefor I made the initialization mechanism work correctly with concurrent invocations.

The `MaxStatementsPerConnection` option was removed because HikariCP doesn't support it.

https://idempiere.atlassian.net/browse/IDEMPIERE-4984

CarlosRuiz-globalqss commented:
Hi, retaking this here ...

I just deployed latest changes on server https://test-feature.idempiere.org:3243/webui/index.zul

There is a new version of HikariCP 5.0.1 with updated dependencies (log4j and h2)

siepkes commented:
> There is a new version of HikariCP 5.0.1 with updated dependencies (log4j and h2)

Log4j is only used by HikariCP in tests but it's indeed probably best to steer clear of the broken versions alltogether to avoid any confusion.

So how do you folks want to proceed with this? Run in test for a while to see if nothing breaks?

CarlosRuiz-globalqss commented:
I would like to do some performance tests to compare c3p0 vs hikaricp.  Thinking what's the best way to do it, I guess a single program will not show results, but we need to stress the database with multiple sessions running in parallel.

CarlosRuiz-globalqss commented:
I'm running the unit tests and it's failing:

from console:
mvn clean verify -DskipTests=false

from eclipse:
run launcher idempiere.unit.test

in both cases there are multiple failures in different test cases with the error:

for example in eclipse:
```
23:29:18.745===========> MSequence.getNextID: AD_PInstance - DB.getConnectionRO - @NoDBConnection@ [1]
java.lang.IllegalStateException: DB.getConnectionRO - @NoDBConnection@
	at org.compiere.util.DB.createConnection(DB.java:477)
	at org.compiere.util.DB.getConnectionID(DB.java:386)
	at org.compiere.model.MSequence.getNextID(MSequence.java:187)
	at org.compiere.util.DB.getNextID(DB.java:1821)
	at org.compiere.model.PO.saveNew(PO.java:3043)
	at org.compiere.model.PO.save(PO.java:2354)
	at org.compiere.process.ServerProcessCtl.process(ServerProcessCtl.java:104)
	at org.compiere.process.ServerProcessCtl.process(ServerProcessCtl.java:58)
	at org.compiere.wf.MWorkflow.runDocumentActionWorkflow(MWorkflow.java:1079)
	at org.idempiere.test.model.AllocationTest.completeDocument(AllocationTest.java:609)
	at org.idempiere.test.model.AllocationTest.testAllocatePaymentPosting(AllocationTest.java:467)
```

for example in console:
```
23:17:22.224===========> MSequence.getNextID: AD_PInstance - DB.getConnectionRO - @NoDBConnection@ [1]
java.lang.IllegalStateException: DB.getConnectionRO - @NoDBConnection@
	at org.compiere.util.DB.createConnection(DB.java:477)
	at org.compiere.util.DB.getConnectionID(DB.java:386)
	at org.compiere.model.MSequence.getNextID(MSequence.java:187)
	at org.compiere.util.DB.getNextID(DB.java:1821)
	at org.compiere.model.PO.saveNew(PO.java:3043)
	at org.compiere.model.PO.save(PO.java:2354)
	at org.compiere.process.ServerProcessCtl.process(ServerProcessCtl.java:104)
	at org.compiere.process.ServerProcessCtl.process(ServerProcessCtl.java:58)
	at org.compiere.wf.MWorkflow.runDocumentActionWorkflow(MWorkflow.java:1079)
	at org.idempiere.test.base.MatchPOTest.testReversePartialMatchPO(MatchPOTest.java:717)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
```

siepkes commented:
Didn't know a test suite existed, good to know! I'll dive in to it.

siepkes commented:
Hi folks! Just a quick ping, I'm still interested in this but I'm currently a bit swamped. I'll try and get back to you as soon as I can!

siepkes commented:
Sorry for the delay but I'm determined to "bring this home"! 

I'm trying to get the integration tests to pass unfortunately I cannot even get the integration tests to pass in a build of `master` without any changes (9ed8c950f7f0f0f6a297c2 at the time of writing). 

I am using OpenJDK 11 and PostgreSQL 13 on Fedora 36 and performed the steps with a clean repo and database. It will fail on `BOMDropFormTest`:

```
testOptimisticLocking  Time elapsed: 0.097 s
testVirtualColumnLoad  Time elapsed: 0.004 s  <<< FAILURE!
org.opentest4j.AssertionFailedError: Wrong value returned ==> expected: <123.45> but was: <0.00>
	at org.idempiere.test.base.POTest.testVirtualColumnLoad(POTest.java:463)

test_Changed  Time elapsed: 0.019 s
testForUpdate  Time elapsed: 5.012 s
Running org.idempiere.test.performance.CacheTest
Tests run: 5, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 1.519 s - in org.idempiere.test.performance.CacheTest
testPrintFormatCacheReset  Time elapsed: 0.557 s
testZoomConditionCache  Time elapsed: 0.009 s
testPOCacheAfterUpdate  Time elapsed: 0.627 s
testTableCache  Time elapsed: 0.015 s
testServices  Time elapsed: 0.311 s
Running org.idempiere.test.form.ChargeFormTest
Tests run: 2, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.049 s - in org.idempiere.test.form.ChargeFormTest
testCreateChargeFromAccount  Time elapsed: 0.027 s
testCreateNewAccontAndCharge  Time elapsed: 0.021 s
Running org.idempiere.test.form.ArchiveViewerFormTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.058 s - in org.idempiere.test.form.ArchiveViewerFormTest
testQuery  Time elapsed: 0.057 s
Running org.idempiere.test.form.BOMDropFormTest
10:23:21.398-----------> MOrderLine.save: beforeSave - MOrderLine[0, Line=0, Ordered=1.00, Delivered=0.0, Invoiced=0.0, Reserved=0.0, LineNet=0.0] [1]
org.adempiere.exceptions.ProductNotOnPriceListException: Product is not on PriceList - Product:Seat, Price List:Standard, Date:07/14/2022
	at org.compiere.model.MOrderLine.beforeSave(MOrderLine.java:838)
	at org.compiere.model.PO.save(PO.java:2264)
	at org.compiere.model.PO.saveEx(PO.java:2442)
	at org.idempiere.apps.form.BOMDrop.addOrderLine(BOMDrop.java:208)
	at org.idempiere.apps.form.BOMDrop.saveOrderLines(BOMDrop.java:189)
	at org.idempiere.test.form.BOMDropFormTest.testDropToOrder(BOMDropFormTest.java:118)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:725)
	at org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:131)
	at org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:149)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestableMethod(TimeoutExtension.java:140)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestMethod(TimeoutExtension.java:84)
	at org.junit.jupiter.engine.execution.ExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(ExecutableInvoker.java:115)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.lambda$invoke$0(ExecutableInvoker.java:105)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)

Tests run: 3, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 0.183 s <<< FAILURE! - in org.idempiere.test.form.BOMDropFormTest
testDropToOrder  Time elapsed: 0.043 s  <<< ERROR!
org.adempiere.exceptions.AdempiereException: Error: : Product is not on PriceList - Product:Seat, Price List:Standard, Date:07/14/2022
	at org.idempiere.test.form.BOMDropFormTest.testDropToOrder(BOMDropFormTest.java:118)
Caused by: org.adempiere.exceptions.ProductNotOnPriceListException: @ProductNotOnPriceList@ - @M_Product_ID@:Seat, @M_PriceList_ID@:Standard, @Date@:07/14/2022
	at org.idempiere.test.form.BOMDropFormTest.testDropToOrder(BOMDropFormTest.java:118)

testDropToInvoice  Time elapsed: 0.088 s
testDropToProject  Time elapsed: 0.052 s

```

My steps are:

```
$ git clone https://github.com/idempiere/idempiere.git
````

Create the database:

```
$ export PGPASSWORD=$(cat $HOME/melissan-postgres-pw)
$ psql -h 127.0.0.1 -p 15432 -U postgres
postgres=# CREATE ROLE adempiere WITH LOGIN ENCRYPTED PASSWORD '<INSERT PASSWORD GENERATED WITH 'pwgen --secure 24'>'; 
postgres=# CREATE SCHEMA adempiere AUTHORIZATION adempiere;
postgres=# GRANT CREATE ON DATABASE postgres TO adempiere;
postgres=# CREATE EXTENSION IF NOT EXISTS "uuid-ossp" SCHEMA public;
```

Load the database seed:

```
$  cd /tmp
$ jar xvf ~/git/idempiere/org.idempiere.p2/target/products/org.adempiere.server.product/linux/gtk/x86_64/data/seed/Adempiere_pg.jar
$ psql -h 127.0.0.1 -p 15432 -d postgres -U adempiere -f Adempiere_pg.dmp -W
```

Configure Eclipse and run the `install.app` to configure iDempiere.

Update the database:

```
$ cd idempiere
./RUN_SyncDBDev.sh
```

Perform the build with integration tests:
```
$ mvn clean verify -DskipTests=false
```

The integration tests will then fail on the above test.

hengsin commented:
hi @siepkes , I've fix the unit test error above, please try again.

siepkes commented:
Thanks @hengsin for the quick fix!

I've rebased the PR on master and removed a unused variable in the PostgreSQL DB implementation. Next I ran the test suite 2 times on this PR via the Maven CLI app and 2 times via Eclipse and it completes normally with PostgreSQL. 

I don't have a working Oracle setup but I'm not expecting any trouble from it. Do you folks have a working Oracle setup on which to run the test suite on this PR?

hengsin commented:
hi @siepkes , lets wait for @CarlosRuiz-globalqss  to test it on Oracle.

hengsin commented on org.compiere.db.oracle.provider/src/org/compiere/db/DB_Oracle.java:
```diff
@@ -319,12 +313,14 @@ public String toString()
+
+            logBuffer.append("# Connections: ").append(mxBean.getTotalConnections());
+            logBuffer.append(" , # Busy Connections: ").append(mxBean.getActiveConnections());
+            logBuffer.append(" , # Idle Connections: ").append(mxBean.getActiveConnections());
```
why getActiveConnections() for both busy and idle ?

siepkes replied:
Good catch! Copy paste error. I resolved it (in multiple places).

hengsin replied:
seems the fix is gone in latest push

hengsin commented on org.compiere.db.oracle.provider/src/org/compiere/db/DB_Oracle.java:
```diff
@@ -570,288 +567,64 @@ public String[] getCommands (int cmdType)
-            throw exception;
-        return conn;
-    }   //  getCachedConnection
+		return conn;
```
this is probably over simplification. we do want to log the exception and keep the "org.adempiere.db.traceStatus" feature.

siepkes replied:
I've re-added the trace function.

The exception is already thrown so why also log it here? Throwing and logging is usually seen as an anti-pattern. It leads to multiple log entries for the same cause since the exception is also going to get logged somewhere higher up. And if higher up it is also "thrown-and-logged" the problem gets even bigger. It makes troubleshooting harder.

Most of the code which I removed was removed because didn't think it was needed anymore. For example the random sleep to wait for a free connection (`int randomNum = rand.nextInt(40 - 20 + 1) + 20;` etc.) is not needed with HikariCP because it will simply block until a connection is available. If you want threads to wait longer for a connection to free up then it's simply a matter of increasing the timeout. And for example the `Runtime.getRuntime().runFinalization();` part was removed because it only hints the GC to call the finalizers _next_ GC cycle. It doesn't free up anything immediately. So the benefit will be minimal.

hengsin commented on org.compiere.db.oracle.provider/src/org/compiere/db/DB_Oracle.java:
```diff
@@ -570,288 +567,64 @@ public String[] getCommands (int cmdType)
 	{
-		//
 		String base = null;
-		if (Ini.isClient())
```
we have deprecated the swing client and I think it is time to gradually remove all this Ini.isClient() check from source.

siepkes replied:
I've removed the `if (Ini.isClient())` in both the PostgreSQL and Oracle DB code.

hengsin replied:
What needs to remain is the Ini.isClient() == false part (i.e the else part here).

siepkes replied:
Right, I need more coffee apparantly ;-). Fixed!

hengsin commented on org.compiere.db.oracle.provider/src/org/compiere/db/DB_Oracle.java:
```diff
@@ -881,36 +656,119 @@ public Connection getDriverConnection (String dbUrl, String dbUid, String dbPwd)
+		{
+			propertyFileURL = Ini.isClient()
+					? OracleBundleActivator.bundleContext.getBundle().getEntry("META-INF/pool/client.default.properties")
+							: OracleBundleActivator.bundleContext.getBundle().getEntry("META-INF/pool/server.default.properties");						
```
Please remove isClient check and always load from server.default.properties

siepkes replied:
Done! I've also removed the actual client properties file.

hengsin replied:
seems all the fix for Ini.IsClient and client.properties is lost in the latest push

hengsin commented on org.compiere.db.postgresql.provider/src/org/compiere/db/DB_PostgreSQL.java:
```diff
@@ -891,40 +680,119 @@ public Connection getDriverConnection (String dbUrl, String dbUid, String dbPwd)
+		{
+			propertyFileURL = Ini.isClient()
+					? PostgreSQLBundleActivator.bundleContext.getBundle().getEntry("META-INF/pool/client.default.properties")
+							: PostgreSQLBundleActivator.bundleContext.getBundle().getEntry("META-INF/pool/server.default.properties");						
```
Please remove isClient check and always load from server.default.properties

siepkes replied:
Done! I've also removed the actual client properties file.

siepkes commented:
I've updated the PR. Thanks for taking the time to look at it!

CarlosRuiz-globalqss commented:
Hi @hengsin  and @siepkes.

I ran the unit tests in postgresql -> success

Tried to run the tests in oracle in a clean vanilla oracle seed -> failed, it created a DB lock very soon and stopped working, the log of mvn shows this:
[mvn_test_oracle.txt](https://github.com/idempiere/idempiere/files/9126288/mvn_test_oracle.txt)

DBeaver shows the table that is holding the lock is C_PROJECT, but I'm not sure I'm interpreting correctly the situation.

After that I tried to run the tests from eclipse, and it created a lock too, and the log showed this error:
```
19:40:43.199===========> DB.executeUpdate: INSERT INTO C_UOM_Conversion (C_UOM_Conversion_ID,AD_Client_ID,AD_Org_ID,IsActive,Created,CreatedBy,Updated,UpdatedBy,C_UOM_ID,C_UOM_To_ID,MultiplyRate,DivideRate,C_UOM_Conversion_UU) VALUES (?,?,?,?,?,?,?,?,?,?,?,?,?) [POSaveC_UOM_Conversion_99c1b8c7-6722-4184-a132-06bfe7201a8b] [1]
java.sql.SQLIntegrityConstraintViolationException: ORA-00001: restricción única (IDEMPIEREDEV.C_UOM_CONVERSION_PRODUCT) violada
; State=23000; ErrorCode=1
```
[eclipse_test_oracle_after.txt](https://github.com/idempiere/idempiere/files/9126292/eclipse_test_oracle_after.txt)

The error is thrown in MUOMConversionTest.testConversion

And the lock is created in MStorageOnHandTest.testGetAll
The lock here appears to be in M_LOT

So, it seems oracle is somehow committing some of the results?

siepkes commented:
Just to be sure; You are not encountering this issue with an unmodified build of `master` and Oracle?

CarlosRuiz-globalqss commented:
> Just to be sure; You are not encountering this issue with an unmodified build of `master` and Oracle?

That's right - testing oracle with the actual core is having the same issue, so, I think we have never "unit tested" in Oracle before, now we need to fix that   :-)

hengsin commented:
ticket and pr created for oracle unit test errors - https://idempiere.atlassian.net/browse/IDEMPIERE-5359

CarlosRuiz-globalqss commented:
OK - ran again the unit tests with oracle + HikariCP and it worked fine.

Anyways, I would like to conduct some performance and stress tests before going ahead with this.

Ideas about how to stress/test it?

siepkes commented:
I think a stress test would involve low-level testing methods such as `getDataSource` for correct behavior under load. I think testing more high level functionality (creating orders and such) is more likely to uncover problems not in the database implementation but in the order creation process and such.

CarlosRuiz-globalqss commented:
I added comments about stress tests in the ticket
https://idempiere.atlassian.net/browse/IDEMPIERE-5013?focusedCommentId=48594

hengsin commented on org.compiere.db.postgresql.provider/src/org/compiere/db/DB_PostgreSQL.java:
```diff
@@ -591,105 +582,51 @@ public RowSet getRowSet (java.sql.ResultSet rs) throws SQLException
+		}
 		
 		//
 		return base + getName() + File.separator + POOL_PROPERTIES;
```
I think it is probably better to rename this to hikaricp.properties to avoid misreading existing pool.properties created for c3p0.

siepkes replied:
Done.

hengsin commented on org.compiere.db.postgresql.provider/src/org/compiere/db/DB_PostgreSQL.java:
```diff
@@ -891,40 +680,117 @@ public Connection getDriverConnection (String dbUrl, String dbUid, String dbPwd)
+	            m_ds.setDriverClassName(DRIVER);
+	            
+	            for(String propertyKey : poolProperties.stringPropertyNames()) {
+	            	m_ds.addDataSourceProperty(propertyKey, poolProperties.get(propertyKey));
```
this is not working, only cachePrepStmts, prepStmtCacheSize and prepStmtCacheSqlLimit is dataSourceProperty, the other are connection pool properties.

[db_postgresql.patch.txt](https://github.com/idempiere/idempiere/files/9275872/db_postgresql.patch.txt)

I use this patch in my local testing but perhaps there are better approach.

Also, note that due to the way logging is currently configure in iDempiere, if you use HikariConfig instead, any invalid property will freeze the system (stuck in the logger).

siepkes replied:
I've changed the way the pool is initialized and the settings files are now loaded correctly.

hengsin commented on org.compiere.db.oracle.provider/META-INF/pool/server.default.properties:
```diff
@@ -1,18 +1,76 @@
+
+cachePrepStmts=true
+prepStmtCacheSize=250
+prepStmtCacheSqlLimit=2048
```
isn't Hikari CP recommend against the use of prepare statement cache ?

siepkes replied:
I assume you are referring to [this section](https://github.com/brettwooldridge/HikariCP/blob/ef99317811093299858d27b54be7b40c43224901/README.md#statement-cache) in HikariCP's readme?

The way I read it is that HikariCP doesn't have a prepared statement cache (because they indeed don't recommend it). However they do recommend to use the JDBC / DB's native statement cache. If I understand it correctly the `prepStmtCacheSqlLimit` is a setting used by the datasource's JDBC driver.

Though something is still wrong here since I think these settings are MySQL specific and don't do anything for PostgreSQL. I'll remove them.

CarlosRuiz-globalqss commented:
Good, after @hengsin patch I'm getting more consistent results, will document in the ticket.

I'm applying the following patch here (based on hengsin's patch, so we just need to apply this one to add both):
[pr926_add.txt](https://github.com/idempiere/idempiere/files/9277296/pr926_add.txt)

Regards,

Carlos Ruiz

hengsin commented:
[926_hikariconfig.patch.txt](https://github.com/idempiere/idempiere/files/9289071/926_hikariconfig.patch.txt)
- use HikariConfig for loading of pool properties
- fix infinite loop when there's invalid pool property in the pool property file

hengsin commented:
hi @siepkes , are you still working on this ?

siepkes commented:
Yes I am! I've reserved some days at the beginning of next week to do some work on this.

CarlosRuiz-globalqss commented:
Hi @siepkes - there is a pending patch also here:
https://github.com/idempiere/idempiere/pull/926#issuecomment-1207463870

hengsin commented:
hi @siepkes , the change for `org.adempiere.base/src/org/compiere/util/CLogErrorBuffer.java` have not been merged in the latest commit. Also, many of the earliest fix have been reverted in the latest commit.

siepkes commented:
Hi @hengsin ! I was still working on it (and more important hadn't tested it yet) which is why I hadn't pinged you folks yet.

I'm still kind of finding the "right" way to be able to collaborate within a PR. Currently I rebase and squish all the changes in to a single commit in order to keep track of what the patch actually looks like. When you merge with master (instead of rebasing) you loose track of all the changes which end up somewhere in between all the changes that got merged in the history. 

I've adapted the `CLogErrorBuffer.java` change you added a bit because in `master`. `CLogErrorBuffer.java` has changed with commit 288e96fe8d472b946ee2cd5238631cacbd9d6fd8.

Concrete I've merged / adapted the all the fixes I "extracted" from the patches:

* [`db_postgresql.patch.txt`] The `Trx.getActiveTransactions()` to `Trx.getOpenTransactions()` was already incorporated with a previous commit.
* [`db_postgresql.patch.txt`] The `System.out.println(Thread.currentThread().getName()+" - "+getStatus());` was added to the `getCachedConnection` method.
* [`db_postgresql.patch.txt`] The hardcoded properties (such as `connectionTimeout`) have been replaced by a way of reading all properties dynamically from the `client.default.properties` and `server.default.properties` files. I've tested this and this now works.
* [`pr926_add.txt`] `202208072022_IDEMPIERE-5013.sql` for Oracle and PostgreSQL were added.
** [`pr926_add.txt`] The corresponding `MSequence` and `MSysConfig` changes were added.
* [`pr926_add.txt`] The changes to the default settings in `server.default.properties` have been added.
* [`pr926_add.txt`] (duplicate) The `System.out.println(Thread.currentThread().getName()+" - "+getStatus());` was added to the `getCachedConnection` method.
* [`pr926_add.txt`] (duplicate) The `Trx.getActiveTransactions()` to `Trx.getOpenTransactions()` was already incorporated with a previous commit.
* [`pr926_add.txt`] (duplicate)The hardcoded properties (such as `connectionTimeout`) have been replaced by a way of reading all properties dynamically from the `client.default.properties` and `server.default.properties` files. I've tested this and this now works.
* [`926_hikariconfig.patch2.txt`] The `CLogErrorBuffer` changes have been added.
* [`926_hikariconfig.patch2.txt`] (duplicate) The changes to the default settings in `server.default.properties` have been added.
* [`926_hikariconfig.patch2.txt`] (duplicate) The `Trx.getActiveTransactions()` to `Trx.getOpenTransactions()` was already incorporated with a previous commit.
* [`926_hikariconfig.patch2.txt`] (duplicate) The `System.out.println(Thread.currentThread().getName()+" - "+getStatus());` was added to the `getCachedConnection` method.


> Also, many of the earlier fix have been reverted in the latest commit.

To which fixes are you refering? Are you refering to the way the HikariCP properties are loaded? I've replaced to changes with a change which doesn't require hardcoding the values.

hengsin commented:
hi @siepkes ,

1. The fix for using mxBean.getActiveConnections() for "# Idle Connections" have been reverted.
2. In earlier commit, you have removed client.default.properties and all the use of Ini.IsClient(). That have been reverted as well.

Regards,
Low

CarlosRuiz-globalqss commented:
Hi @siepkes - it is easier to review if you add commits to the pull request instead of replacing it completely.

That way we can review just the changes on the last commits that we have not reviewed yet.

siepkes commented:
@CarlosRuiz-globalqss Yes you are right. I'm not really used with working with "long running" Github PR's. I usually work with Gerrit or Phabricator.

I just noticed some change fell out. I'll fix it by doing a separate commit.

siepkes commented:
Alright, fixed.
